### ISSUE TITLE
Executables in TRUSTED_PLATFORM_ASSEMBLIES prevents finding any subsequent dlls

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/runtime#62735]
### Description

The problem occurs with Xamarin.Mac apps, when:

1. `System.Private.CoreLib.dll` is *not* next to libcoreclr.dylib
2. We use `TRUSTED_PLATFORM_ASSEMBLIES` to tell CoreCLR where to find `System.Private.CoreLib.dll`.
3. An executable (`*.exe`) is listed before `System.Private.CoreLib.dll` in `TRUSTED_PLATFORM_ASSEMBLIES`.

Looking at the code, this is what I see:

1. CoreCLR first tries to load System.Private.CoreLib.dll from the directory where libcoreclr.dylib is:

https://github.com/dotnet/runtime/blob/3992fc3841133e15ad322842d35adb8c249a824f/src/coreclr/binder/assemblybinder.cpp#L381-L387

2. When that fails, CoreCLR looks at `TRUSTED_PLATFORM_ASSEMBLIES`:

https://github.com/dotnet/runtime/blob/3992fc3841133e15ad322842d35adb8c249a824f/src/coreclr/binder/assemblybinder.cpp#L393

iterating over all the paths therein:

https://github.com/dotnet/runtime/blob/3992fc3841133e15ad322842d35adb8c249a824f/src/coreclr/binder/assemblybinder.cpp#L397

and exiting the loop if something goes wrong:

https://github.com/dotnet/runtime/blob/3992fc3841133e15ad322842d35adb8c249a824f/src/coreclr/binder/assemblybinder.cpp#L403-L407

the problem is that `dllOnly=true` is passed to the `GetNextTPAPath` function, and in that case, the `GetNextTPAPath` function returns a failure if it finds anything but a dll (such as an .exe):

https://github.com/dotnet/runtime/blob/3992fc3841133e15ad322842d35adb8c249a824f/src/coreclr/binder/utils.cpp#L246-L247

I tried looking for any docs specifying whether the entries in `TRUSTED_PLATFORM_ASSEMBLIES` need to be sorted in any way, but I couldn't find any mention of it.

### Reproduction Steps

I tried creating a repro from a simple console app, but that didn't work (didn't look like `TRUSTED_PLATFORM_ASSEMBLIES` is even looked at, but I didn't investigate much).

I can create a rather complex test project with a Xamarin.Mac app, but that seems a bit overkill when I believe someone knowledgeable about CoreCLR should be able to create a simple test case pretty fast.

### Expected behavior

The order of assemblies in `TRUSTED_PLATFORM_ASSEMBLIES` shouldn't matter.

### Actual behavior

If exes come before dlls, then those dlls might not be found.

### Known Workarounds

Put dlls before exes in `TRUSTED_PLATFORM_ASSEMBLIES`.

### Configuration

.NET 6 (currently the 6.0-maui branch) + Xamarin.Mac

**Comment by agocke** [comment]
This is currently by-design. `System.Private.CoreLib.dll` must live next to coreclr, it is not intended to be configurable via `TRUSTED_PLATFORM_ASSEMBLIES`.

**Comment by vitek-karas** [comment]
I agree that SDK will always put CoreLib next to coreclr. But the runtime has code to be able to get the CoreLib path from TPA - and apparently that code gets confused if there's an .exe in TPA as well. Currently SDK should never get us into this situation, but we want other scenarios to work as well.
I think it' would be worth it to look into fixing this - but the priority is relatively low (unless this causes problems for Xamarin?)

Just curious: How did the `.exe` get there? Especially on macOS...

**Comment by rolfbjarne** [comment]
When we build a universal app bundle, we have two different `System.Private.CoreLib.dll` assembly files (one for `osx-x64` and one for `osx-arm64`), and a single, universal `libcoreclr.dylib`.

We can't put both versions of `System.Private.CoreLib.dll` in the same directory, so we put them in a subdirectory, and use `TRUSTED_PLATFORM_ASSEMBLIES` to tell CoreCLR where to find it depending on the architecture at runtime (x64 or arm64).

**Comment by rolfbjarne** [comment]
> Just curious: How did the `.exe` get there? Especially on macOS...

Why wouldn't it?

We add all the assemblies in the app bundle to `TRUSTED_PLATFORM_ASSEMBLIES`, and .exes are assemblies.

**Comment by rolfbjarne** [comment]
> > and .exes are assemblies.
> 
> Well - yes, but technically only .NET Framework should produce them. That's why I'm curious - if this is something specific to some app, or potentially common.

I don't remember the exact scenario anymore.

I was able to work around it fairly easy though (make sure any .exes come last in `TRUSTED_PLATFORM_ASSEMBLIES`), so it's not all that important for us.

**Comment by agocke** [comment]
> When we build a universal app bundle,

I see, if this is for universal binary support that might up the priority.

**Comment by rolfbjarne** [comment]
I don't think this is needed after https://github.com/dotnet/runtime/pull/128734 (which we're using in .NET 11).
